src/transformStateWithClones.js

```diff
@@ -7,7 +7,31 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const copyState = { ...state };
+  const history = [];
+
+  for (const act of actions) {
+    switch (act.type) {
+      case 'addProperties':
+        Object.assign(copyState, act.extraData);
+        break;
+
+      case 'removeProperties':
+        for (const key of act.keysToRemove) {
+          delete copyState[key];
+        }
+        break;
+
+      case 'clear':
+        for (const key in copyState) {
+          delete copyState[key];
+        }
+        break;
```

Comment:
Checklist: [CODE STYLE]: switch/case should always have default case for error handling.